api/undertaker/undertaker.go

```diff
@@ -0,0 +1,132 @@
+package undertaker
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/api/common"
+	"github.com/juju/juju/apiserver/params"
+)
+
+// Client provides access to the undertaker api
+type Client struct {
+	base.ClientFacade
+	st     api.Connection
+	facade base.FacadeCaller
+}
+
+// UnderTakerClient defines the methods on the undertaker API end point.
+type UndertakerClient interface {
+	ProcessDyingEnviron() error
+	RemoveAllEnvironDocs() error
+	Life() (params.Life, error)
+	IsStateServer() (bool, error)
+	EnvironUUID() (string, error)
+	TimeOfDeath() (time.Time, error)
+}
```

Comment:
It occurs to me that a lot of the above are quite strange things to expose at the API.

What if more of the undertaker's logic was implemented in state and the API server under a single call named, AdvanceDyingEnviron or something? If the env is dying it could do what ProcessDyingEnviron currently does, and if the env is dead it could handle the document deletion after the ripTime. It would return a bool to indicate when things were done (so that the undertaker could stop).

This would simplify things and avoid some of these awkward API definitions which expose somewhat internal concerns.